crates/prices/src/core/estimate.rs

```diff
@@ -9,79 +10,86 @@ use {
+    fn estimate(&self, swap: Swap) -> BoxFuture<'_, Result<(Price, eth::Gas), EstimatorError>>;
+}
+
 /// Estimate the outcome of a [`Swap`].
 pub async fn estimate(
     swap: Swap,
-    deadline: Deadline,
+    gas_price: eth::GasPrice,
     estimators: &[Box<dyn Estimator>],
 ) -> Result<Estimate, Error> {
-    // Fetch the estimate for the swap.
-    let estimate = median_estimate(swap, deadline, estimators).await?;
-
-    // Convert the fee from ETH to [`ToToken`].
-    let fee = median_estimate(
-        Swap {
-            from: eth::ETH_TOKEN.into(),
-            to: swap.to,
-            amount: estimate.gas.0.into(),
-        },
-        deadline,
-        estimators,
+    let ((swap_price, gas), (eth_price, _)) = try_join(
+        // Fetch the estimate for the swap price.
+        median_estimate(swap, estimators),
+        // Fetch the estimate for ETH price, to convert the gas fees into [`ToToken`].
+        median_estimate(
+            Swap {
+                // TODO Automatically wrap ETH, and specify the chain ID in the config
+                // from: eth::ETH_TOKEN.into(),
+                from: eth::H160::from_str("0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2")
+                    .unwrap()
+                    .into(),
+                to: swap.to,
+                amount: eth::U256::from(1000000000000000u64).into(),
+            },
+            estimators,
+        ),
     )
     .await?;
 
     Ok(Estimate {
-        amount: estimate.amount,
-        fee: fee.amount,
+        amount: swap_price * swap.amount,
+        fee: eth_price * swap::FromAmount::from(gas_price * gas),
     })
 }
 
-/// Fetch the median [`estimator::Estimate`] from the [`Estimator`]s.
+/// Fetch the median swap price and gas fee estimate from the [`Estimator`]s.
 async fn median_estimate(
     swap: Swap,
-    deadline: Deadline,
     estimators: &[Box<dyn Estimator>],
-) -> Result<estimator::Estimate, Error> {
+) -> Result<(Price, eth::Gas), Error> {
     // Fetch all estimates from the estimators.
-    let mut estimates: Vec<_> = join_all(
-        estimators
-            .iter()
-            .map(|estimator| estimator.estimate(swap, deadline)),
-    )
-    .await
-    .into_iter()
-    .filter_map(|estimate| {
-        // TODO Observe the error
-        estimate.ok()
-    })
-    .collect();
+    let (mut prices, mut gas_fees): (Vec<_>, Vec<_>) =
+        join_all(estimators.iter().map(|estimator| estimator.estimate(swap)))
+            .await
+            .into_iter()
+            .filter_map(|estimate| {
+                // TODO Observe the error
+                estimate.ok()
+            })
+            .unzip();
 
-    // Pick the median estimate.
-    estimates.sort_by_key(|estimate| estimate.amount);
-    estimates.get(estimates.len() / 2).copied().ok_or(Error)
+    // Pick the median estimated price and gas fee.
+    prices.sort_unstable_by_key(|&price| price * swap.amount);
+    gas_fees.sort_unstable();
```

Comment:
I think we should not pick the median `price` independently from the median `gas_fees` because they relate to the same trade route and are therefore not independent from each other.

Reply:
Yeah, I had second thoughts about this as well. I think you're right.